src/sources/kubernetes_logs/mod.rs

```diff
@@ -0,0 +1,277 @@
+    transforms::Transform,
+};
+use evmap10::{self as evmap};
+use file_source::{FileServer, FileServerShutdown, Fingerprinter};
+use futures::{future::FutureExt, sink::Sink, stream::StreamExt};
+use futures01::sync::mpsc;
+use k8s_openapi::api::core::v1::Pod;
+use k8s_paths_provider::K8sPathsProvider;
+use pod_metadata_annotator::PodMetadataAnnotator;
+use serde::{Deserialize, Serialize};
+use std::path::PathBuf;
+use std::time::Duration;
+use tokio::task::spawn_blocking;
+
+/// The key we use for `file` field.
+const FILE_KEY: &str = "file";
+
+/// Configuration for the `kubernetes_logs` source.
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields, default)]
+pub struct Config {
+    /// The `name` of the Kubernetes `Node` that Vector runs at.
+    /// Required to filter the `Pod`s to only include the ones with the log
+    /// files accessible locally.
+    self_node_name: String,
+
+    /// Automatically merge partial events.
+    #[serde(default = "crate::serde::default_true")]
+    auto_partial_merge: bool,
+
+    /// Specifies the field names for metadata annotation.
+    annotation_fields: pod_metadata_annotator::FieldsSpec,
+}
+
+inventory::submit! {
+    SourceDescription::new_without_default::<Config>(COMPONENT_NAME)
+}
+
+const COMPONENT_NAME: &str = "kubernetes_logs";
+
+#[typetag::serde(name = "kubernetes_logs")]
+impl SourceConfig for Config {
+    fn build(
+        &self,
+        name: &str,
+        globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<sources::Source> {
+        // TODO: switch to the ones provided by the topology once PRs for that
+        // pass.
+        let rt = crate::runtime::Runtime::new()?;
+        let exec = rt.executor();
+        let resolver = Resolver::new(globals.dns_servers.clone(), exec)?;
+
+        let source = Source::init(self, resolver, globals, name)?;
+
+        // TODO: this is a workaround for the legacy futures 0.1.
+        // When the core is updated to futures 0.3 this should be simplied
+        // significantly.
+        let out = futures::compat::Compat01As03Sink::new(out);
+        let fut = source.run(out, shutdown);
+        let fut = fut.map(|result| {
+            result.map_err(|error| {
+                error!(message = "source future failed", ?error);
+            })
+        });
+        let fut = Box::pin(fut);
+        let fut = futures::compat::Compat::new(fut);
+        let fut: sources::Source = Box::new(fut);
+        Ok(fut)
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        COMPONENT_NAME
+    }
+}
+
+#[derive(Clone)]
+struct Source {
+    client: k8s::client::Client,
+    self_node_name: String,
+    data_dir: PathBuf,
+    auto_partial_merge: bool,
+    fields_spec: pod_metadata_annotator::FieldsSpec,
+}
+
+impl Source {
+    fn init(
```

Comment:
Seems like this would be conventionally called `new`?

Reply:
Good point! :smile: